src/serve_data.js

```diff
@@ -162,6 +165,146 @@ export const serve_data = {
+          }
+
+          //console.log(tileCenter + " ## " + xy);
+          item.source.getTile(z, xy[0], xy[1], async (err, data, headers) => {
+            if (err) {
+              if (/does not exist/.test(err.message)) {
+                return res.status(204).send();
+              } else {
+                return res
+                  .status(500)
+                  .header('Content-Type', 'text/plain')
+                  .send(err.message);
+              }
+            } else {
+              if (data == null) {
+                return res.status(404).send('Not found');
+              } else {
+                if (tileJSONFormat === 'pbf') {
+                  return res.status(404).send('Invalid format');
+                }
+                var image = new Image();
+                image.onload = function () {
+                  const imgSize = 256;
+                  var canvas = createCanvas(imgSize, imgSize);
+
+                  var context = canvas.getContext('2d');
+                  context.drawImage(image, 0, 0);
+
+                  var imgdata = context.getImageData(0, 0, imgSize, imgSize);
+
+                  var index = (xy[1] * imgdata.width + xy[0]) * 4;
+                  var red = imgdata.data[index];
+                  var green = imgdata.data[index + 1];
+                  var blue = imgdata.data[index + 2];
+                  let elevation;
+                  if (item.source._info.encoding == 'mapbox') {
+                    elevation =
+                      -10000 + (red * 256 * 256 + green * 256 + blue) * 0.1;
+                  } else if (item.source._info.encoding == 'terrarium') {
+                    elevation = red * 256 + green + blue / 256 - 32768;
+                  } else {
+                    elevation = 'invalid encoding';
+                  }
+                  //"index": index, "length": imgdata.data.length,
+                  return res.status(200).send({
+                    z: z,
+                    x: xy[0],
+                    y: xy[1],
+                    red: red,
+                    green: green,
+                    blue: blue,
+                    latitude: tileCenter[0],
+                    longitude: tileCenter[1],
+                    elevation: elevation,
+                  });
+                };
+                image.onerror = (err) => {
+                  return res
+                    .status(500)
+                    .header('Content-Type', 'text/plain')
+                    .send(err.message);
+                };
+                //image.onerror = err => { return res.status(200).header('Content-Type', 'image/webp').send(data); }
+
+                if (item.source._info.format == 'webp') {
+                  const img = await sharp(data).toFormat('png').toBuffer();
+                  image.src = img;
+                } else {
+                  image.src = data;
+                }
+              }
+            }
+          });
+        }
+      },
+    );
+
```

Comment:
I suggest this change to add in pmtiles support. pmtiles is nice for raster-dems since it has deduplication built in.
```suggestion
    app.get(
      '^/:id/elevation/:z([0-9]+)/:x([-.0-9]+)/:y([-.0-9]+)',
      async (req, res, next) => {
        try {
          const item = repo?.[req.params.id];
          if (!item) return res.sendStatus(404);

          const { source, tileJSON, sourceType } = item;

          if (!source) return res.status(404).send('Missing source');

          const encoding = tileJSON?.encoding;
          const format = tileJSON?.format;

          if (!encoding) return res.status(404).send('Missing source.encoding');

          if (encoding !== 'terrarium' && encoding !== 'mapbox') {
            return res
              .status(404)
              .send('Missing encoding. Encoding must be terrarium or mapbox.');
          }
          if (!format) return res.status(404).send('Missing format.');

          if (format !== 'webp' && format !== 'png') {
            return res
              .status(404)
              .send('Wrong Image format. Image format must be webp or png.');
          }

          if (!tileJSON) return res.status(404).send('Missing tileJSON.');

          if (!tileJSON.format)
            return res.status(404).send('Missing tileJSON format.');

          if (!sourceType) return res.status(404).send('Missing sourceType.');

          if (sourceType !== 'pmtiles' && sourceType !== 'mbtiles') {
            return res
              .status(404)
              .send('Invalid format. Format must be pmtiles or mbtiles.');
          }

          const z = parseInt(req.params.z, 10);
          let x = req.params.x;
          let y = req.params.y;
          const isInt = /^-?[0-9]+$/.test(x) && /^-?[0-9]+$/.test(y);

          if (
            typeof tileJSON.minzoom === 'undefined' ||
            typeof tileJSON.maxzoom === 'undefined'
          ) {
            return res.status(404).send(JSON.stringify(tileJSON));
          }

          let tileCenter;
          let xy;

          if (isInt) {
            x = parseInt(x, 10);
            y = parseInt(y, 10);

            if (
              z < tileJSON.minzoom ||
              x < 0 ||
              y < 0 ||
              z > tileJSON.maxzoom ||
              x >= Math.pow(2, z) ||
              y >= Math.pow(2, z)
            ) {
              return res.status(404).send('Out of bounds');
            }

            xy = [x, y];
            tileCenter = new SphericalMercator().bbox(x, y, z);
          } else {
            x = parseFloat(x);
            y = parseFloat(y);

            if (
              z < tileJSON.minzoom ||
              x < -180 ||
              y < -90 ||
              z > tileJSON.maxzoom ||
              x > 180 ||
              y > 90
            ) {
              return res.status(404).send('Out of bounds');
            }

            tileCenter = [y, x, y + 0.1, x + 0.1];
            const { minX, minY } = new SphericalMercator().xyz(tileCenter, z);
            xy = [minX, minY];
          }

          let data;
          if (sourceType === 'pmtiles') {
            const tileinfo = await getPMtilesTile(source, z, x, y);
            if (!tileinfo?.data) return res.status(204).send();
            data = tileinfo.data;
          } else {
            data = await new Promise((resolve, reject) => {
              source.getTile(z, xy[0], xy[1], (err, tileData) => {
                if (err) {
                  return /does not exist/.test(err.message)
                    ? resolve(null)
                    : reject(err);
                }
                resolve(tileData);
              });
            });
          }
          if (data === null) return res.status(204).send();
          if (!data) return res.status(404).send('Not found');

          const image = new Image();
          await new Promise((resolve, reject) => {
            image.onload = () => {
              const imgSize = 256;
              const canvas = createCanvas(imgSize, imgSize);
              const context = canvas.getContext('2d');
              context.drawImage(image, 0, 0);
              const imgdata = context.getImageData(0, 0, imgSize, imgSize);

              const arrayWidth = imgdata.width;
              const arrayHeight = imgdata.height;
              const bytesPerPixel = 4;

              // Validate xy values first
              const xPixel = Math.floor(xy[0]);
              const yPixel = Math.floor(xy[1]);

              if (
                xPixel < 0 ||
                yPixel < 0 ||
                xPixel >= arrayWidth ||
                yPixel >= arrayHeight
              ) {
                return reject('Out of bounds Pixel');
              }

              const index = (yPixel * arrayWidth + xPixel) * bytesPerPixel;
              const red = imgdata.data[index];
              const green = imgdata.data[index + 1];
              const blue = imgdata.data[index + 2];

              let elevation;
              if (encoding === 'mapbox') {
                elevation =
                  -10000 + (red * 256 * 256 + green * 256 + blue) * 0.1;
              } else if (encoding === 'terrarium') {
                elevation = red * 256 + green + blue / 256 - 32768;
              } else {
                elevation = 'invalid encoding';
              }

              resolve(
                res.status(200).send({
                  z,
                  x: xy[0],
                  y: xy[1],
                  red,
                  green,
                  blue,
                  latitude: tileCenter[0],
                  longitude: tileCenter[1],
                  elevation,
                }),
              );
            };

            image.onerror = (err) => reject(err);

            if (format === 'webp') {
              sharp(data)
                .toFormat('png')
                .toBuffer()
                .then((img) => {
                  image.src = img;
                })
                .catch(reject);
            } else {
              image.src = data;
            }
          });
        } catch (err) {
          return res
            .status(500)
            .header('Content-Type', 'text/plain')
            .send(err.message);
        }
      },
    );

```